src/services/onchain-data.service.ts

```diff
@@ -69,8 +71,14 @@ export class OnchainDataService implements OnModuleInit
             const token1Address: string = ethers.utils.getAddress((await pairContract.method(token1ABI).call())
                 .decoded[0]);
 
-            const token0: IToken = await this.fetchToken(token0Address);
-            const token1: IToken = await this.fetchToken(token1Address);
+            const token0: IToken = await this.mutex.runExclusive(() =>
+            {
+                return this.fetchToken(token0Address);
+            });
+            const token1: IToken = await this.mutex.runExclusive(() =>
+            {
+                return this.fetchToken(token1Address);
+            });
```

Comment:
This feels unsafe, can these calls not be run in parallel? What if caller 1 gets mutex on token1 and caller 2 gets mutex on token2?

Reply:
Nope they cannot be run in parallel. The mutex blocks it until the function runs to completion. Personally verified it many times

Reply:
Thats just because of ordering of the callbacks on the event loop... Feels a bit unsafe especially between node versions. E.g. if both calls hit the token0 murex one will run the other will wait but who's to say which will hit the token1 mutex first?

Reply:
It doesn't matter. There is only one mutex, not two. So whoever gets the mutex first will go and fetch the token info. The other 'thread' will wait till fetchToken finishes

Reply:
Is the mutex not released after the first promise completes?

![image](https://user-images.githubusercontent.com/11343499/144538067-3c0638e9-60e1-49d1-8bd3-d484cecf42d6.png)

Reply:
Correct. After 1 completes, the line below or some other thread will grab it and run

Reply:
So there is no concern with one thread getting mutexfor token1 then other thread for token2?